### PR TITLE
Use Moto account ID resolver

### PR DESCRIPTION
**Pull request by viren-nadkarni**
### Summary

- Replace all references to Account ID with Moto resolver introduced in https://github.com/spulec/moto/pull/5098
- Use subscripting to retrieve Moto backend from BackendDict so that a new backend is created for the given region if it doesn't exist
- Change internal call detection logic to make use of a newly introduced header. Previously we relied on a special value for AWS_ACCESS_KEY_ID. With the multi-accounts revamp, AWS_ACCESS_KEY_ID must always be set to the Account ID

### Tests

CI pipeline against the ext branch: https://github.com/localstack/localstack/actions/runs/2429784539

### CODE REVIEW

**Review by whummer — COMMENTED** [review]
Awesome, this is a massive set of changes, 🚀 big step towards proper account handling and multi-account support - kudos @viren-nadkarni ! 🎉  Did an initial review for now - have a few questions/suggestions.. Happy to discuss in more detail!

File: localstack/services/events/provider.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import time
 from typing import Any, Dict, List, Optional
 
+from moto.core import get_account_id
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/events/provider.py:10]: this is also the most important change for me!

File: localstack/constants.py [hunk 2]
```diff
@@ -159,6 +161,11 @@
 INTERNAL_AWS_ACCESS_KEY_ID = "__internal_call__"
 INTERNAL_AWS_SECRET_ACCESS_KEY = "__internal_call__"
 
+# This header must be set to the AWS Account ID
+# Presence of this header in an incoming request typically means that the request originated within localstack,
+# i.e. it is an internal cross-service call.
+HEADER_LOCALSTACK_ACCOUNT_ID = "x-localstack-account-id"
```
Comment (thrau) [inline localstack/constants.py:167]: love it!

File: localstack/services/awslambda/lambda_api.py [hunk 3]
```diff
@@ -77,11 +78,12 @@
 from localstack.utils.http import canonicalize_headers, parse_chunked_data
 from localstack.utils.patch import patch
 
-# logger
 LOG = logging.getLogger(__name__)
 
 # name pattern of IAM policies associated with Lambda functions (name/qualifier)
 LAMBDA_POLICY_NAME_PATTERN = "lambda_policy_{name}_{qualifier}"
+LAMBDA_TEST_ROLE = "arn:aws:iam::%s:role/lambda-test-role" % get_account_id()
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/awslambda/lambda_api.py:85]: i also think we should not have code in there that resolve the account id on module level, to not give anyone else who sees the code crazy ideas ;-)

File: localstack/utils/aws/aws_stack.py [hunk 4]
```diff
@@ -877,15 +883,16 @@ def kinesis_stream_name(kinesis_arn):
...
+) -> dict[str, str]:
     ctype = APPLICATION_AMZ_JSON_1_0
     if service == "kinesis":
         ctype = APPLICATION_AMZ_JSON_1_1
     elif service in ["sns", "sqs", "sts", "cloudformation"]:
         ctype = APPLICATION_X_WWW_FORM_URLENCODED
 
-    # TODO: consider adding an internal=False flag, to use INTERNAL_AWS_ACCESS_KEY_ID for internal calls here
-    access_key = access_key or constants.TEST_AWS_ACCESS_KEY_ID
+    access_key = access_key or get_account_id()
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/utils/aws/aws_stack.py:895]: i was also going to add the same comment elsewhere. i think we should probably not replace instances of `TEST_AWS_ACCESS_KEY_ID`, but rather have the account id resolver map the `TEST_AWS_ACCESS_KEY_ID` to the account id from the current context.

File: tests/integration/terraform/provider.tf [hunk 5]
```diff
@@ -4,7 +4,7 @@ variable "region_name" {
 
 provider "aws" {
   region                      = var.region_name
-  access_key                  = "test"
+  access_key                  = "000000000000"
```
[the remark this answers is not in the document]
Comment (thrau) [reply tests/integration/terraform/provider.tf:7]: +1

**Review by thrau — CHANGES_REQUESTED** [review]
Awesome changes @viren-nadkarni, thanks for grinding through them!

My high-level comments are as Waldemar's
* let's introduce our own **internal abstraction for getting the account id**. i suggest that we implement the same pattern as we introduced in https://github.com/spulec/moto/pull/5098
  - introduce an `account_id_resolver` method that can be patched
  - use that method in our own `get_account_id` method
  - patch `moto.core.account_id_resolver` with our `get_account_id` method

  note how this does not actually wrap moto but rather changes the moto behavior (inversion of control)

* do not by default set the access key id in requests to the account id. rather map requests that have `INTERNAL_AWS_ACCESS_KEY_ID` in their auth info to whatever the `account_id_resolver` returns.

**Comment by viren-nadkarni** [comment]
> but rather have the account id resolver map the TEST_AWS_ACCESS_KEY_ID to the account id from the current context.

@thrau Is it possible to have variables mapped to functions? Closest is the `@property` decorator but that's not applicable here.

File: tests/integration/terraform/provider.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (viren-nadkarni) [reply tests/integration/terraform/provider.tf:7]: We now have a basic translation layer which maps the access key `test` to account ID `000000000000`.

This should ensure backward compatilbity and also have no impact on presigned S3 URLS.

File: localstack/utils/aws/aws_stack.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (viren-nadkarni) [reply localstack/utils/aws/aws_stack.py]: with [881ce200](https://github.com/localstack/localstack/commit/881ce200e2f7d7d0ef22328ab75340f337e5fc43#diff-c249821a753f15353e8eaf413c9109fee4c2cfcab9df57c6fbf2fb9022270039R900), we now use the original access key ID from the request wherever available

**Review by thrau — APPROVED** [review]
This looks very clean now @viren-nadkarni, great job!

Just one minor nit: it would be great if we could move `accounts.py` into `localstack.aws` and also move the newly introduced hook for patching moto into `accounts.py`.

File: localstack/aws/accounts.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    return _TEST_AWS_ACCOUNT_ID
+
+
+account_id_resolver = get_default_account_id
+
+
+def get_account_id_from_access_key_id(access_key_id: str) -> str:
+    """Return the Account ID associated the Access Key ID."""
+    # This utility ignores IAM mappings.
+    # For now, we assume the client sends Account ID in Access Key ID field.
```
[the remark this answers is not in the document]
Comment (viren-nadkarni) [reply localstack/aws/accounts.py:28]: Good catch. Fixed in da0ce0ff

**Pull request merged**